dotnet/src/Connectors/Connectors.AI.Oobabooga/TextCompletion/OobaboogaTextCompletion.cs

```diff
@@ -0,0 +1,296 @@
+                            break;
+                        default:
+                            break;
+                    }
+                }
+                else if (result.MessageType == WebSocketMessageType.Close)
+                {
+                    await clientWebSocket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, "Acknowledge Close frame", CancellationToken.None).ConfigureAwait(false);
+                }
+
+                if (clientWebSocket.State != WebSocketState.Open)
+                {
+                    break;
+                }
+            }
+        }
+        finally
+        {
+            s_activeConnections.TryPop(out _);
+
+            if (clientWebSocket != null && !isClientProvidedWebSocket)
+            {
+                if (s_activeConnections.IsEmpty)
+                {
+                    while (s_webSocketPool.TryPop(out ClientWebSocket clientToDispose))
+                    {
+                        clientToDispose.Dispose();
+                    }
+
+                    clientWebSocket.Dispose();
+                }
+                else
+                {
+                    s_webSocketPool.Push(clientWebSocket);
+                }
+            }
+        }
+    }
+
+    /// <inheritdoc/>
+    public async Task<IReadOnlyList<ITextResult>> GetCompletionsAsync(
+        string text,
+        CompleteRequestSettings requestSettings,
+        CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            var blockingUri = new UriBuilder(this._endpoint)
+            {
+                Port = this._blockingPort,
+                Path = BlockingUriPath
+            };
+
+            var completionRequest = this.CreateOobaboogaRequest(text, requestSettings);
+
+            using var stringContent = new StringContent(
+                JsonSerializer.Serialize(completionRequest),
+                Encoding.UTF8,
+                "application/json");
+
+            using var httpRequestMessage = new HttpRequestMessage()
+            {
+                Method = HttpMethod.Post,
+                RequestUri = blockingUri.Uri,
+                Content = stringContent
+            };
+            httpRequestMessage.Headers.Add("User-Agent", HttpUserAgent);
+
+            using var response = await this._httpClient.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
+            response.EnsureSuccessStatusCode();
+
+            var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+
+            TextCompletionResponse? completionResponse = JsonSerializer.Deserialize<TextCompletionResponse>(body);
+
+            if (completionResponse is null)
+            {
+                throw new AIException(AIException.ErrorCodes.InvalidResponseContent, "Unexpected response from model")
```

Comment:
When implementing the exception, please wrap it in a dedicated exception where you expose the details in a property.

Please see:
[dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs](https://github.com/microsoft/semantic-kernel/blob/1efe78aab4472460b71895efc85b65506a6f177b/dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/ClientBase.cs#L62C23-L62C53) and [dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/OpenAIInvalidResponseException{T}.cs#L9](https://github.com/microsoft/semantic-kernel/blob/1efe78aab4472460b71895efc85b65506a6f177b/dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/OpenAIInvalidResponseException%7BT%7D.cs#L9)

Reply:
@RogerBarreto the new exception type is [there](https://github.com/microsoft/semantic-kernel/pull/1357/files#diff-99b9245cb239afb1fe54f86541c7c96cf34b0f3aa52d7c87fc08d41c42119af4R8).